granary/flickr.py

```diff
@@ -0,0 +1,371 @@
+
+import appengine_config
+import datetime
+import functools
+import itertools
+import json
+import logging
+import requests
+import source
+import sys
+import mf2py
+import urllib2
+
+
+class Flickr(source.Source):
+
+  DOMAIN = 'flickr.com'
+  NAME = 'Flickr'
+
+  API_EXTRAS = ','.join(('date_upload', 'date_taken', 'views', 'media',
+                         'description' 'tags', 'machine_tags', 'geo'))
+
+  def __init__(self, access_token_key, access_token_secret, user_id=None):
+    """Constructor.
+
+    Args:
+      access_token_key: string, OAuth access token key
+      access_token_secret: string, OAuth access token secret
+      user_id: string, the logged in user's Flickr nsid (optional)
+    """
+    self.access_token_key = access_token_key
+    self.access_token_secret = access_token_secret
+    self.user_id = user_id
+
+  def call_api_method(self, method, params={}):
+    """Call a Flickr API method.
+    """
+    return flickr_auth.call_api_method(
+      method, params, self.access_token_key, self.access_token_secret)
+
+  def get_activities_response(self, user_id=None, group_id=None, app_id=None,
+                              activity_id=None, start_index=0, count=0,
+                              etag=None, min_id=None, cache=None,
+                              fetch_replies=False, fetch_likes=False,
+                              fetch_shares=False, fetch_events=False,
+                              search_query=None):
+    """Get Flickr actvities
+    """
+    if user_id is None:
+      user_id = 'me'
+    if group_id is None:
+      group_id = source.FRIENDS
+
+    params = {}
+    method = None
+    solo = False
+
+    if activity_id:
+      method = 'flickr.photos.getInfo'
+      params['photo_id'] = activity_id
+      solo = True
+    elif group_id == source.SELF:
+      method = 'flickr.people.getPhotos'
+      params['user_id'] = user_id
+      params['extras'] = self.API_EXTRAS
+      params['per_page'] = 50
+    elif group_id == source.FRIENDS:
+      method = 'flickr.photos.getContactsPhotos'
+      params['extras'] = self.API_EXTRAS
+      params['per_page'] = 50
+    elif group_id == source.ALL:
+      method = 'flickr.photos.getRecent'
+      params['extras'] = self.API_EXTRAS
+      params['per_page'] = 50
+
+    assert method
+    logging.debug('calling %s with %s', method, params)
+    photos_resp = self.call_api_method(method, params)
+
+    result = {'items': []}
+    if solo:
+      photos = [photos_resp.get('photo', {})]
+    else:
+      photos = photos_resp.get('photos', {}).get('photo', [])
+
+    for photo in photos:
+      activity = self.photo_to_activity(photo)
+
+      # TODO consider using 'flickr.activity.userPhotos' when group_id=@self,
+      # gives all recent comments and faves, instead of hitting the API for
+      # each photo
+      if fetch_replies:
+        replies = []
+        comments_resp = self.call_api_method('flickr.photos.comments.getList', {
```

Comment:
sad that we can't use async urlfetch since granary isn't app engine specific, nor can we take the liberty of spawning threads (or whatever) to do parallel HTTP requests.

current status: trying to find any async support in requests. one halfway measure would be to use `requests.Session` to at least reuse a single connection.

not necessary for this PR, i'm fine with just adding a TODO for whatever we settle on.